pkg/machine/compression/gzip.go

```diff
@@ -0,0 +1,39 @@
+package compression
+
+import (
+	"io"
+
+	image "github.com/containers/image/v5/pkg/compression"
+	"github.com/sirupsen/logrus"
+)
+
+type gzipDecompressor struct {
+	genericDecompressor
+}
+
+func newGzipDecompressor(compressedFilePath string) (*gzipDecompressor, error) {
+	d, err := newGenericDecompressor(compressedFilePath)
+	return &gzipDecompressor{*d}, err
+}
+
+func (d *gzipDecompressor) decompress(w WriteSeekCloser, r io.Reader) error {
+	gzReader, err := image.GzipDecompressor(r)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := gzReader.Close(); err != nil {
+			logrus.Errorf("Unable to close gz file: %q", err)
+		}
+	}()
+
+	sparseWriter := NewSparseWriter(w)
+	defer func() {
+		if err := sparseWriter.Close(); err != nil {
```

Comment:
This error should be actually returned to the caller; it makes no sense to proceed if the file is incomplete. (In readers, `Close` failing is pretty unlikely, so, meh, logging is fine.)

In all three instances — probably best handled in a shared utility.